rpc/client/client.go

```diff
@@ -500,7 +508,63 @@ func (c *Client) TxSearch(ctx context.Context, query string, prove bool, pagePtr
 // BlockSearch defines a method to search for a paginated set of blocks by
 // BeginBlock and EndBlock event search criteria.
 func (c *Client) BlockSearch(ctx context.Context, query string, page, perPage *int, orderBy string) (*ctypes.ResultBlockSearch, error) {
-	panic("BlockSearch - not implemented!")
+	q, err := tmquery.New(query)
+	if err != nil {
+		return nil, err
+	}
+
+	results, err := c.node.BlockIndexer.Search(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort the results
+	switch orderBy {
+	case "desc":
+		sort.Slice(results, func(i, j int) bool {
+			return results[i] > results[j]
+		})
+
+	case "asc", "":
+		sort.Slice(results, func(i, j int) bool {
+			return results[i] < results[j]
+		})
+	default:
+		return nil, errors.New("expected order_by to be either `asc` or `desc` or empty")
+	}
+
+	// Paginate
+	totalCount := len(results)
+	perPageVal := validatePerPage(perPage)
+
+	pageVal, err := validatePage(page, perPageVal, totalCount)
+	if err != nil {
+		return nil, err
+	}
+
+	skipCount := validateSkipCount(pageVal, perPageVal)
+	pageSize := tmmath.MinInt(perPageVal, totalCount-skipCount)
+
+	pageResults := make([]int64, 0, pageSize)
+	copy(pageResults, results[skipCount:skipCount+pageSize])
+
+	// Fetch the blocks
+	blocks := make([]*ctypes.ResultBlock, 0, len(pageResults))
+	for _, h := range pageResults {
+		b, err := c.node.Store.LoadBlock(uint64(h))
+		if err != nil {
+			return nil, err
+		}
+		block, err := abciconv.ToABCIBlock(b)
+		if err != nil {
+			return nil, err
+		}
+		blocks = append(blocks, &ctypes.ResultBlock{
+			Block: block,
```

Comment:
`BlockID` is not filled. Optimint doesn't use `PartSetHeader`, but `Hash` should be filled.